umapi_client/connection.py

```diff
@@ -310,6 +313,20 @@ def execute_multiple(self, actions, immediate=True):
         for a in actions:
             if len(a.commands) == 0:
                 if self.logger: self.logger.warning("Sending action with no commands: %s", a.frame)
+            # throttling part 1a: split commands with group assignments > self.throttle_groups
+            # if an action gets more than 10 commands per verb/group type combination, then it will
+            # be split below
+            for i, c in enumerate(a.commands):
+                for verb in ['add', 'remove']:
+                    subcommand = c.get(verb, {})
+                    if not subcommand:
+                        continue
+                    for group_type in ['product', 'usergroup', 'productConfiguration']:
+                        groups = subcommand.get(group_type, [])
+                        if not groups:
+                            continue
+                        if len(groups) > self.throttle_groups:
```

Comment:
This code looks like it makes a separate command for each list that's too long, which is both inefficient and logically incorrect (because it doesn't preserve order of operations).  When you integrate this testing with the actual code that does the splitting, you should do the same logic that is done in the throttling code below: duplicate the dictionary of groups, truncate all group lists in the original to only have 10, remove the first 10 from the all the group lists in the copy, and then remove any keys in the copy with empty group lists.  If the mutated copy is not empty, make a single additional command (immediately after the original in the action list, to preserve ordering), and then iterate over that new command (in case the group lists were really long).

Reply:
"immediately after the original in the action list, to preserve ordering" -- you mean command list, right?  Because my intent is to let the existing action splitting logic work if my command splitter adds too many commands to an individual action.

Reply:
Yes the commands list, sorry for the typo.